# Experiment 7 - Text Analysis

Chirag Rana - 2018130043

In [2]:
import pandas as pd 
import numpy as np
# !pip install nltk
from nltk.stem import PorterStemmer
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.ensemble import RandomForestClassifier

### Problem 1.1 – Loading the Dataset

In [3]:
df = pd.read_csv('emails.csv')

In [4]:
df.head()

text  spam
0  Subject: naturally irresistible your corporate...     1
1  Subject: the stock trading gunslinger  fanny i...     1
2  Subject: unbelievable new homes made easy  im ...     1
3  Subject: 4 color printing special  request add...     1
4  Subject: do not have money , get software cds ...     1

In [5]:
df.describe()

spam
count  5728.000000
mean      0.238827
std       0.426404
min       0.000000
25%       0.000000
50%       0.000000
75%       0.000000
max       1.000000

In [6]:
df.shape

(5728, 2)

In [7]:
spam_count = np.sum(np.asarray(df.spam) ) 
spam_count

1368

In [8]:
# Check of 1st word:
max_length = 0
for item in np.asarray(df.text):
    max_length = max(len(item), max_length)
    texts = item.split(":")
    if texts[0] != "Subject": 
        print("Subject is not the First word")
        break 

print("Maximum length of the longest text is: ", max_length)
    

Maximum length of the longest text is:  43952


### Answers of Problem 1

* How many emails are in the dataset?<br>
-> There are 5728 emails in the Dataset.


* How many of the emails are spam?<br>
-> Count of Spam emails = 1368

* Which word appears at the beginning of every email in the dataset? <br>
"Subject" is the first word in all the texts. <br>

* Respond as a lower-case word with punctuation removed. <br>
"subject"

* Could a spam classifier potentially benefit from including the frequency of the word that appears in every email?<br>
-> Yes -- the number of times the word appears might help us differentiate spam from ham.

* The nchar() function counts the number of characters in a piece of text. <br>
How many characters are in the longest email in the dataset (where longest is measured in terms of the maximum number of characters)?<br>
-> 43952


### Problem 2.1 -Preparing the Corpus

In [9]:
corpus = df.copy()

In [10]:
corpus['text'] = corpus['text'].str.replace('[^a-zA-Z#]', " ")
corpus

C:\Users\Dexter\AppData\Local\Temp\ipykernel_8768\3801576629.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  corpus['text'] = corpus['text'].str.replace('[^a-zA-Z#]', " ")


text  spam
0     Subject  naturally irresistible your corporate...     1
1     Subject  the stock trading gunslinger  fanny i...     1
2     Subject  unbelievable new homes made easy  im ...     1
3     Subject    color printing special  request add...     1
4     Subject  do not have money   get software cds ...     1
...                                                 ...   ...
5723  Subject  re   research and development charges...     0
5724  Subject  re   receipts from visit  jim    than...     0
5725  Subject  re   enron case study update  wow   a...     0
5726  Subject  re   interest  david    please   call...     0
5727  Subject  news   aurora       update  aurora ve...     0

[5728 rows x 2 columns]

In [11]:
corpus['text']
corpus_arr = list(map(lambda x: x.lower(), corpus['text']))
len(corpus_arr)

5728

In [12]:
stop_words = ["","i", "me", "my", "myself", "we", "our", "ours", "ourselves", "you", "your", "yours", "yourself", "yourselves", "he", "him", "his", "himself", "she", "her", "hers", "herself", "it", "its", "itself", "they", "them", "their", "theirs", "themselves", "what", "which", "who", "whom", "this", "that", "these", "those", "am", "is", "are", "was", "were", "be", "been", "being", "have", "has", "had", "having", "do", "does", "did", "doing", "would", "should", "could", "ought", "i'm", "you're", "he's", "she's", "it's", "we're", "they're", "i've", "you've", "we've", "they've", "i'd", "you'd", "he'd", "she'd", "we'd", "they'd", "i'll", "you'll", "he'll", "she'll", "we'll", "they'll", "isn't", "aren't", "wasn't", "weren't", "hasn't", "haven't", "hadn't", "doesn't", "don't", "didn't", "won't", "wouldn't", "shan't", "shouldn't", "can't", "cannot", "couldn't", "mustn't", "let's", "that's", "who's", "what's", "here's", "there's", "when's", "where's", "why's", "how's", "a", "an", "the", "and", "but", "if", "or", "because", "as", "until", "while", "of", "at", "by", "for", "with", "about", "against", "between", "into", "through", "during", "before", "after", "above", "below", "to", "from", "up", "down", "in", "out", "on", "off", "over", "under", "again", "further", "then", "once", "here", "there", "when", "where", "why", "how", "all", "any", "both", "each", "few", "more", "most", "other", "some", "such", "no", "nor", "not", "only", "own", "same", "so", "than", "too", "very"]
corpus_text = np.asarray(corpus_arr)

In [13]:
corpus_without_sw = []
for sentence in corpus_text:
  corpus_without_sw.append(' '.join([word for word in sentence.split() if not word in stop_words]))
corpus_without_sw

['subject naturally irresistible corporate identity lt really hard recollect company market full suqgestions information isoverwhelminq good catchy logo stylish statlonery outstanding website will make task much easier promise havinq ordered iogo company will automaticaily become world ieader isguite ciear without good products effective business organization practicable aim will hotat nowadays market promise marketing efforts will become much effective list clear benefits creativeness hand made original logos specially done reflect distinctive company image convenience logo stationery provided formats easy use content management system letsyou change website content even structure promptness will see logo drafts within three business days affordability marketing break shouldn t make gaps budget satisfaction guaranteed provide unlimited amount changes extra fees surethat will love result collaboration look portfolio interested',
 'subject stock trading gunslinger fanny merrill muzo col

In [14]:
corpus = corpus_without_sw
len(corpus)

5728

##  Stemmming of Words

In [15]:
porter = PorterStemmer()
corpus_stemmed = []
for sentence in corpus:
  corpus_stemmed.append(' '.join([porter.stem(word) for word in sentence.split()]))

corpus_stemmed

['subject natur irresist corpor ident lt realli hard recollect compani market full suqgest inform isoverwhelminq good catchi logo stylish statloneri outstand websit will make task much easier promis havinq order iogo compani will automaticaili becom world ieader isguit ciear without good product effect busi organ practic aim will hotat nowaday market promis market effort will becom much effect list clear benefit creativ hand made origin logo special done reflect distinct compani imag conveni logo stationeri provid format easi use content manag system letsyou chang websit content even structur prompt will see logo draft within three busi day afford market break shouldn t make gap budget satisfact guarante provid unlimit amount chang extra fee surethat will love result collabor look portfolio interest',
 'subject stock trade gunsling fanni merril muzo colza attaind penultim like esmark perspicu rambl segovia group tri slung kansa tanzania ye chameleon continu clothesman libretto chesapea

In [16]:
corpus = corpus_stemmed
len(corpus)

5728

In [17]:
df['texts'] = corpus

# Building Document Term Matrix

In [18]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(df['texts'], df['spam'], test_size=0.33, random_state=42)

In [19]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfvec = TfidfVectorizer()
X_train = tfvec.fit_transform(X_train)
X_test = tfvec.transform(X_test)

In [20]:
X_train.shape

(3837, 21070)

In [21]:
X_test.shape

(1891, 21070)

In [22]:
dtm = pd.DataFrame(X_train.toarray(),columns = tfvec.get_feature_names())
dtm

C:\Users\Dexter\Anaconda3\envs\mp_project\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


aa  aaa  aagraw  aal  aaldou  aall  aanalysi  aaron   ab  aba  ...  \
0     0.0  0.0     0.0  0.0     0.0   0.0       0.0    0.0  0.0  0.0  ...   
1     0.0  0.0     0.0  0.0     0.0   0.0       0.0    0.0  0.0  0.0  ...   
2     0.0  0.0     0.0  0.0     0.0   0.0       0.0    0.0  0.0  0.0  ...   
3     0.0  0.0     0.0  0.0     0.0   0.0       0.0    0.0  0.0  0.0  ...   
4     0.0  0.0     0.0  0.0     0.0   0.0       0.0    0.0  0.0  0.0  ...   
...   ...  ...     ...  ...     ...   ...       ...    ...  ...  ...  ...   
3832  0.0  0.0     0.0  0.0     0.0   0.0       0.0    0.0  0.0  0.0  ...   
3833  0.0  0.0     0.0  0.0     0.0   0.0       0.0    0.0  0.0  0.0  ...   
3834  0.0  0.0     0.0  0.0     0.0   0.0       0.0    0.0  0.0  0.0  ...   
3835  0.0  0.0     0.0  0.0     0.0   0.0       0.0    0.0  0.0  0.0  ...   
3836  0.0  0.0     0.0  0.0     0.0   0.0       0.0    0.0  0.0  0.0  ...   

      zwrocic  zwwyw  zwzm  zxghlajf  zyc  zygoma  zymg  zzn  zzncacst  zzzz  
0         0.0    0.0   0.0       0.0  0.0     0.0   0.0  0.0       0.0   0.0  
1         0.0    0.0   0.0       0.0  0.0     0.0   0.0  0.0       0.0   0.0  
2         0.0    0.0   0.0       0.0  0.0     0.0   0.0  0.0       0.0   0.0  
3         0.0    0.0   0.0       0.0  0.0     0.0   0.0  0.0       0.0   0.0  
4         0.0    0.0   0.0       0.0  0.0     0.0   0.0  0.0       0.0   0.0  
...       ...    ...   ...       ...  ...     ...   ...  ...       ...   ...  
3832      0.0    0.0   0.0       0.0  0.0     0.0   0.0  0.0       0.0   0.0  
3833      0.0    0.0   0.0       0.0  0.0     0.0   0.0  0.0       0.0   0.0  
3834      0.0    0.0   0.0       0.0  0.0     0.0   0.0  0.0       0.0   0.0  
3835      0.0    0.0   0.0       0.0  0.0     0.0   0.0  0.0       0.0   0.0  
3836      0.0    0.0   0.0       0.0  0.0     0.0   0.0  0.0       0.0   0.0  

[3837 rows x 21070 columns]

## How many terms are in dtm?
### Problem 3.1
### 1) CART 

In [23]:
from sklearn import tree
clf = tree.DecisionTreeClassifier(random_state=111)

In [24]:
clf = clf.fit(X_train, y_train)

In [25]:
clf.score(X_test,y_test)

0.9561078794288737

### 2) Random Forest

In [27]:
scores = []
for i in range(200):
  clf_RF = RandomForestClassifier(oob_score=True,random_state=i)
  clf_RF.fit(X_train, y_train)
  scores.append(clf_RF.score(X_test,y_test))
  print(i,clf_RF.score(X_test,y_test))

0 0.9672131147540983
1 0.9698572184029614
2 0.9666842940243258
3 0.9656266525647805
4 0.9709148598625066
5 0.9666842940243258
6 0.970386039132734
7 0.9698572184029614
8 0.9656266525647805
9 0.9687995769434162
10 0.9661554732945531
11 0.9661554732945531
12 0.9672131147540983
13 0.9661554732945531
14 0.9656266525647805
15 0.9661554732945531
16 0.9661554732945531
17 0.9661554732945531
18 0.9640401903754627
19 0.9682707562136436
20 0.967741935483871
21 0.9629825489159175
22 0.9640401903754627
23 0.965097831835008
24 0.9635113696456901
25 0.9645690111052353
26 0.9693283976731888
27 0.9714436805922793
28 0.967741935483871
29 0.9656266525647805
30 0.9672131147540983
31 0.9687995769434162
32 0.9672131147540983
33 0.9709148598625066
34 0.9698572184029614
35 0.9672131147540983
36 0.9725013220518244
37 0.9672131147540983
38 0.9661554732945531
39 0.9656266525647805
40 0.9682707562136436
41 0.9645690111052353
42 0.9725013220518244
43 0.9656266525647805
44 0.9624537281861449
45 0.967741935483871
46 

In [28]:
print(scores.index(max(scores)),max(scores))

111 0.9740877842411423


### Conclusion:
In this Experiment, I was able to do text analysis by creating a corpus (Making lowercase, removing punctuations, removing stopwords, stemming words). 

I was able to compare the CART model and Random Forest model. There was about a 2% increase in accuracy in the Random Forest (Ensemble model). However, CART model ran very much faster as compared to Random Forest. 
